dashboard/app/helpers/aichat_safety_helper.rb

```
@@ -42,26 +42,37 @@ def find_toxicity(role, text, locale, level_id)
       details = nil
       start_time = Time.now
       report_openai_safety_check("Start")
-      # Try twice in case of network errors or model not correctly following directions and
-      # replying with something other valid expected output.
       attempts = 1
+
       Retryable.retryable(tries: 2) do
```

Were you able to test this out manually? Seems like this could potentially make a total of four requests (ie, the retry with structured output is within this "retryable" block, which is set up to retry)?

---

Oh sorry I think I'm misunderstanding how this retryable thing works, lemme try and wrap my head around it!

---

I think you're still technically correct! I'm actually currently working on a quick reformat so that it only retries on a network error instead; there might be a general discussion to be had on whether it makes sense to possibly allow a reattempt if the fallback call fails due to a network errors (e.g. 1st attempt = fail on unstructured -> network fail on structured (2nd call) -> retry on structured (3rd call)) or if we'd rather just let it fail on the retry to keep it hard locked at 2 calls no matter what.

---

FWIW I'd probably propose dropping the retries wrapper entirely for simplicity's sake since we're basically managing the retry ourself, and users can retry if both our standard safety check and structured output check fails. The remaining "OpenAI failed for some other reason" situation seems pretty rare (I see only a couple instances of "OpenAI failed with status" over the past 6 months): https://app.honeybadger.io/projects/3240/faults/114776339/01JEWCT1DNZFZSQMJHK8C46D4H?q=OpenAI+request+failed+with+status

---

I agree; alternatively if we want to stick to retryable in case we want to bump the number of retries in the future, then it may be clearer to attempt structured output on the "last retry" rather than explicitly on attempt #2. But given our historical data, I agree that dropping retryable altogether might just be simpler.

---

Closing this thread to make room for the thread revisiting this topic after all of the recent changes!